lightning-invoice/src/payment.rs

```diff
@@ -773,25 +738,53 @@ where
 	}
 }
 
+/// A map with liquidity value (in msat) keyed by a short channel id and the direction the HTLC
+/// is traveling in. The direction boolean is determined by checking if the HTLC source's public
+/// key is less than its destination. See [`InFlightHtlcs::used_liquidity_msat`] for more
+/// details.
+pub struct InFlightHtlcs(HashMap<(u64, bool), u64>);
+
+impl InFlightHtlcs {
+	/// Returns liquidity in msat given the public key of the HTLC source, target, and short channel
+	/// id.
+	pub fn used_liquidity_msat(&self, source: &NodeId, target: &NodeId, channel_scid: u64) -> Option<&u64> {
+		self.0.get(&(channel_scid, source < target))
+	}
+}
+
+#[cfg(not(feature = "no-std"))]
```

Comment:
Why does this have to be `std`?

Reply:
Ah, I guess we have to do something like this:

https://github.com/lightningdevkit/rust-lightning/blob/48d21bad7b0b8dc567ed0e4e3f55c62e48fcb422/lightning/src/lib.rs#L84-L89

I didn't realize at the time that we already have a `core2` dependency in this crate.

Reply:
Mh, couldn't we just use the re-export `lightning::io`? That's what I recently did in #1708 .

Reply:
Yeah, that would work. I just wasn't sure what our conventions were, tbh. We have a dependency in `lightning-invoice`'s `Cargo.toml` file  from https://github.com/lightningdevkit/rust-lightning/commit/01915810d42506addbaef244b0cbb31e8b3856b5 but don't appear to use it anywhere outside of doc tests.

Reply:
SGTM

Reply:
so it sounds like we want something like this instead of only implementing it for std?

```
#[cfg(feature = "std")]
/// Re-export of either `lightning::io` or `std::io`, depending on the `std` feature flag.
pub use std::io;
#[cfg(not(feature = "std"))]
/// Re-export of either `lightning::io` or `std::io`, depending on the `std` feature flag.
pub use lightning::io;

...

impl lightning::util::ser::Writeable for InFlightHtlcs {
	fn write<W: lightning::util::ser::Writer>(&self, writer: &mut W) -> Result<(), io::Error> { self.0.write(writer) }
}

impl lightning::util::ser::Readable for InFlightHtlcs {
	fn read<R: lightning::io::Read>(reader: &mut R) -> Result<Self, lightning::ln::msgs::DecodeError> {
		let infight_map: HashMap<(u64, bool), u64> = lightning::util::ser::Readable::read(reader)?;
		Ok(Self(infight_map))
	}
}

```

Reply:
You should only need `use lightning::io;` (no `pub`, no conditional import). Basically, just inherit which `io` is used from the `lightning` crate rather than doing any configuration in `lightning-invoice`. Also, update doc tests accordingly.

Reply:
`read`'s type parameter should use `io::Read` in its trait bound, too.

Reply:
👍